Deny service ClusterIP update from `None`

nebril opened the pull request:
**What this PR does / why we need it**: Headless service should not be transformed into a service with ClusterIP, therefore update of this field if it's set to `None` is disallowed.

**Which issue this PR fixes** _(optional, in `fixes #<issue number>(, #<issue_number>, ...)` format, will close that issue when PR gets merged)_: fixes #33029

**Release note**:

``` release-note
Changing service ClusterIP from `None` is not allowed anymore.
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34094)

<!-- Reviewable:end -->

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -2575,6 +2575,10 @@ func ValidateServiceUpdate(service, oldService *api.Service) field.ErrorList {
 	}
 
+	if oldService.Spec.ClusterIP == api.ClusterIPNone {
+		allErrs = append(allErrs, ValidateImmutableField(service.Spec.ClusterIP, oldService.Spec.ClusterIP, field.NewPath("spec", "clusterIP"))...)
```
This block is identical to above.  Can we just flatten this to a unilateral `ValidateImmutableField` call?  I can't think of any valid update of this field.

CREATE sets "" -> allocated
UPDATE A -> B is disallowed
UPDATE A -> "" is disallowed
UPDATE A -> "None" is disallowed
UPDATE None -> B is disallowed
UPDATE None -> "" is disallowed
UPDATE None -> "None" is disallowed

We should never every see it as ""

nebril replied:
Makes sense, I will update it.

nebril replied:
@thockin what about the case in https://github.com/kubernetes/kubernetes/blob/caad4db78be6ed/test/e2e/dns.go#L481 ? It seems like you can change ClusterIP when changing type of service. I think my check should prevail.

thockin replied:
oh, good point.

OK, can we then make this code more obvious and comment it heavily to explain which transitions are being caught and which are not?

nebril replied:
Sure thing.

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -2571,7 +2571,9 @@ func validateServicePort(sp *api.ServicePort, requireName, isHeadlessService boo
-	if api.IsServiceIPSet(oldService) {
+	// ClusterIP should be mutable only for empty value, to allow assigning ClusterIP
+	// for service changing type to one that requires ClusterIP
+	if oldService.Spec.ClusterIP != "" {
```
What about the other direction?  We've been a little lax in the definition here.  Do we think it is valid to change from ClusterIP to ExternalName types?

thockin replied:
I wonder if it makes more sense to refactor this into a top-level switch on service.Spec.Type and just handle each case in a more obvious way...  Thoughts?

nebril replied:
Seems like it. Will try to carefully think of all cases.

nebril committed:
Deny ClusterIP update for services using it

If the service is of the type using the ClusterIP (ClusterIP, NodePort,
LoadBalancer), the update of ClusterIP is prohibited.

nebril committed:
Test cases for service ClusterIP updates

Test cases from ClusterIP using types to other ClusterIP using types
(ClusterIP, NodePort, LoadBalancer) added.

thockin commented:
Let me know when ready for review

nebril commented:
@thockin I think it's ready, I covered most of the cases in tests, everything seems to work fine.

nebril commented:
@thockin rebased, pinging for review

nebril commented:
Agreed, IP should be allocated on such change. Will see if there is an issue for that, if not I will create one.

nebril commented:
OK, I found #35354
